### PR TITLE
CI: ICC/ICPC & ICX/ICPX

### PR DESCRIPTION
**Pull request by ax3l**
- [x] Add Intel `icc`/`icpc` from oneAPI to CI.
- [x] Add Intel `icx`/`icpx` from oneAPI to CI.

### CODE REVIEW

File: .github/workflows/unix.yml [hunk 1]
```diff
@@ -188,3 +188,33 @@ jobs:
...
+    - name: Install
+      run: |
+        sudo apt-get -qqq update
+        sudo apt-get install g++ libopenmpi-dev libhdf5-openmpi-dev
+        sudo apt-get install -y wget build-essential pkg-config cmake ca-certificates gnupg
+        sudo wget https://apt.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS-2023.PUB
+        sudo apt-key add GPG-PUB-KEY-INTEL-SW-PRODUCTS-2023.PUB
+        echo "deb https://apt.repos.intel.com/oneapi all main" | sudo tee /etc/apt/sources.list.d/oneAPI.list
+        sudo apt-get update
+        sudo apt-get install -y intel-oneapi-compiler-dpcpp-cpp-and-cpp-classic python3 python3-numpy python3-mpi4py
```
Comment (ax3l) [inline .github/workflows/unix.yml]:
```suggestion
        sudo apt-get install -y intel-oneapi-compiler-dpcpp-cpp-and-cpp-classic intel-oneapi-mpi-devel intel-oneapi-python
```

File: .github/workflows/unix.yml [hunk 2]
```diff
@@ -188,3 +188,33 @@ jobs:
...
         ctest --output-on-failure
+
+  icc_py39_mpi_h5:
+    runs-on: ubuntu-20.04
+    steps:
+    - uses: actions/checkout@v2
+    - name: Install
+      run: |
+        sudo apt-get -qqq update
+        sudo apt-get install g++ libopenmpi-dev libhdf5-openmpi-dev
```
Comment (ax3l) [inline .github/workflows/unix.yml]:
```suggestion
        sudo apt-get install g++ libopenmpi-dev
```

File: .github/workflows/unix.yml [hunk 3]
```diff
@@ -188,3 +188,33 @@ jobs:
         cmake -S .. -B . -DopenPMD_USE_PYTHON=ON -DopenPMD_USE_MPI=ON -DopenPMD_USE_HDF5=ON -DopenPMD_USE_INVASIVE_TESTS=ON
         cmake --build . --parallel 2
         ctest --output-on-failure
+
+  icc_py39_mpi_h5:
```
Comment (ax3l) [inline .github/workflows/unix.yml]:
```suggestion
  icc_py39_mpi:
```

File: .github/workflows/unix.yml [hunk 4]
```diff
@@ -188,3 +188,33 @@ jobs:
...
+      run: |
+        set +e; source /opt/intel/oneapi/setvars.sh; set -e
+        mkdir build && cd build
+        ../share/openPMD/download_samples.sh && chmod u-w samples/git-sample/*.h5
+        cmake -S .. -B . \
+          -DCMAKE_C_COMPILER=$(which icc)    \
+          -DCMAKE_CXX_COMPILER=$(which icpc) \
+          -DopenPMD_USE_PYTHON=ON \
+          -DopenPMD_USE_MPI=ON    \
+          -DopenPMD_USE_HDF5=ON   \
```
Comment (ax3l) [inline .github/workflows/unix.yml:217]:
```suggestion
```

Comment (ax3l) [inline .github/workflows/unix.yml]:
```suggestion
        cmake -S .. -B . \
          -DCMAKE_C_COMPILER=$(which icc)    \
          -DCMAKE_CXX_COMPILER=$(which icpc) \
          -DopenPMD_USE_PYTHON=ON \
          -DopenPMD_USE_MPI=ON    \
```

File: .github/workflows/unix.yml [hunk 5]
```diff
@@ -188,3 +188,31 @@ jobs:
...
+      env: {CXXFLAGS: -Werror -Wno-deprecated-declarations}
+      run: |
+        set +e; source /opt/intel/oneapi/setvars.sh; set -e
+        mkdir build && cd build
+        cmake -S .. -B . \
+          -DCMAKE_C_COMPILER=$(which icc)    \
+          -DCMAKE_CXX_COMPILER=$(which icpc) \
+          -DopenPMD_USE_PYTHON=ON \
+          -DopenPMD_USE_MPI=ON    \
+          -DopenPMD_USE_INVASIVE_TESTS=ON
```
Comment (ax3l) [inline .github/workflows/unix.yml]:
```suggestion
          -DopenPMD_USE_MPI=ON
```

**Comment by ax3l** [comment]
Ready for merge, just depends on fix for ICC (regression of #931)

**Comment by franzpoeschel** [comment]
Hmm, tests are failing now:
```
[~Series] An error occurred: Dynamic cast returned a nullptr!
```
Funny enough, tests run fine when building PR #886 with `icc`, so I think that cleaning up our hierarchy representation really has its merits. I'd propose to wait for #886 before merging this one?
I've opened a demonstration PR #934 that rebases this one onto #886 to show this.

**Comment by ax3l** [comment]
Horray, ICC & ICX coverage added :) Thanks for the fix, @franzpoeschel! :sparkles:

**Pull request merged**